src/theory/quantifiers/global_negate.cpp

```diff
@@ -0,0 +1,112 @@
+/*********************                                                        */
+/*! \file global_negate.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of global_negate
+ **/
+
+#include "theory/quantifiers/global_negate.h"
+#include "theory/rewriter.h"
+
+using namespace std;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+void GlobalNegate::simplify(std::vector<Node>& assertions)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  Assert(!assertions.empty());
+
+  Trace("cbqi-gn") << "Global negate : " << std::endl;
+  // collect free variables in all assertions
+  std::unordered_set<Node, NodeHashFunction> free_vars;
+  for (const Node& as : assertions)
+  {
+    Trace("cbqi-gn") << "  " << as << std::endl;
+    TNode cur = as;
+    // compute free variables
+    std::unordered_set<TNode, TNodeHashFunction> visited;
+    std::unordered_set<TNode, TNodeHashFunction>::iterator it;
+    std::vector<TNode> visit;
```

Comment:
Ditto. Push all assertions onto visit vector (`visit.insert(visit.end(), assertions.begin(), assertions.end())`).

Reply:
I'd prefer keeping the for loop here to keep the structure (and there is a Trace message in the for loop).